ISPN-13446 IRAC: configure the tombstone cleanup task

pruivo opened the pull request:
* Dynamic task based on a target map size: if the current number of stored
  tombstones are higher, decrease the schedule interval to make the
  cleanup task more frequent (and vice-versa).

oraNod reviewed (approved):
I pushed another doc edit @pruivo and LGTM. Thanks!

danberindei reviewed (changes requested):
I can't say I love the algorithm for adjusting the delay...
I put a comment in DefaultIracTombstoneManager with what I think would be a better approach.

pruivo commented on core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java:
```diff
@@ -122,21 +136,29 @@ public void startCleanupTombstone() {
-         stage.dependsOn(new CleanupTask(entry.getKey(), entry.getValue()).checkRemoteSites());
+
+         return stage.freeze().whenComplete((unused, throwable) -> {
+            double tombstoneRatio = tombstoneMap.size() * 1.0 / configuration.tombstoneMapSize();
```
ok, it makes sense. I tried to keep it simpler but computing the creating rate will help to achieve a better interval. Hopefully, the delay will stabilize after a while (assuming the workload does not change).

I'm also going to drop the "DynamicTask" class since everything is calculated here.

pruivo commented:
updated with the following changes
* Changed attribute name from "interval" to "delay".
* Added test to the scheduler suggested by Dan.
* Added metric to report the current delay value.
* DynamicTask removed.

danberindei commented on core/src/main/java/org/infinispan/configuration/parsing/Attribute.java:
```diff
@@ -105,6 +105,7 @@
     MAPPER,
     MARSHALLER,
     MAX_BATCH_SIZE,
+   MAX_CLEANUP_DELAY,
```
Indent

danberindei commented on core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java:
```diff
@@ -220,4 +262,64 @@ public void run() {
+               return;
+            }
+         }
+         double tombstoneCreationRate = (initialSize - previousPostCleanupSize) * 1.0 / currentDelayMillis;
```
`currentDelayMillis` can be 0 here, if size >= targetSize.

I'm starting to think I should have let you use the initial algorithm, as this is getting complicated, but I'll have another stab :)

I see you've covered 2 corner cases that I missed in my previous suggestion:
1. It's impossible to compute the rate of change when the delay is 0.
2. Sometimes the creation rate is negative.

Unfortunately, I don't like your solution, because I don't like "magic constants" like the 100ms delay step :)

My proposal would be to:
1. Make the minimum delay 1 instead of 0. It shouldn't have any visible effect on the speed of cleaning up tombstones, but it would get rid of the division by zero risk.
2. When the creation rate is negative, I believe we should be using the maximum delay.

There is one more thing that you're touching a bit with they way you're handling the negative creation rate: you're making the changes incremental, instead of using just the creation rate from the last interval you're also using the creation rate from the previous interval (via the current delay).

I've given it some thought and I think if it's probably worth doing something similar every time, not just when the creation rate is negative, but I'd rather not use a constant step, instead we could do an average between the next delay and the last delay.

So the algorithm ends up something like this:

if (size >= targetSize) {
   currentDelay = 1;
} else {
tombstoneCreationRate = (initialSize - previousPostCleanupSize) * 1.0 / currentDelayMillis
if (tombstoneCreationRate <= 0) {
  nextDelay = maxDelayMillis;
} else {
  nextDelay = (targetSize - size) / tombstoneCreationRate;
}
currentDelayMillis = (currentDelayMillis + nextDelay) / 2;
}

I'm going to try it out myself, hopefully it won't end up adding even more complications...

I did consider doing something with the time between the start and the end of the cleanup task as well, since tombstones are also being created during that time, but at this point I decided that would be too much, even for me :)

danberindei replied:
Ok, my experiment is at https://github.com/infinispan/infinispan/compare/main...danberindei:pr_pruivo_t_13446

Let me know what you think @pruivo ...

pruivo replied:
looks good to me. I've copied your changes.

danberindei commented on core/src/test/java/org/infinispan/xsite/irac/IracTombstoneUnitTest.java:
```diff
@@ -0,0 +1,265 @@
+      manager.stop();
+   }
+
+   public void testDelayConstantAtHigherRate() throws InterruptedException {
```
@pruivo one more nit to pick: are `testDelayConstantAtHigherRate` and `testDelayConstantAtLowerRate` named correctly? Because the delay doesn't seem to stay constant to me :)

pruivo replied:
updated
